File: ts/src/additive.ts
import {
  type AdditiveRequest,
  type AdditiveTerm,
  type ExplainedComponent,
  type Interval,
  parseRequest,
  parseScoreResult,
  type ScoreResult,
} from "./contracts.js";
import { ContractCode, ContractValidationError } from "./errors.js";
import { inputsHash } from "./requestHash.js";

function finiteOutput(value: number): number {
  if (!Number.isFinite(value)) {
    throw new ContractValidationError(ContractCode.INVALID_NUMBER);
  }
  return value === 0 ? 0 : value;
}

function contribution(term: AdditiveTerm, value = term.value): number {
  return finiteOutput(value * term.coefficient);
}

function termBounds(term: AdditiveTerm): readonly [number, number] {
  if (term.interval === null) {
    const point = contribution(term);
    return [point, point];
  }
  return [
    contribution(term, term.interval.low),
    contribution(term, term.interval.high),
  ];
}

function contributionInterval(term: AdditiveTerm): Interval | null {
  const [low, high] = termBounds(term);
  return low === high ? null : { low, high };
}

function row(term: AdditiveTerm): ExplainedComponent {
  return {
    id: term.id,
    raw: term.value,
    normalized: null,
    declared_weight: null,
    operation: term.operation,
    coefficient: term.coefficient,
    contribution: contribution(term),
    contribution_interval: contributionInterval(term),
  };
}

function apply(
  total: number,
  amount: number,
  operation: "add" | "subtract",
): number {
  return finiteOutput(operation === "add" ? total + amount : total - amount);
}

function final(value: number, request: AdditiveRequest): number {
  const result = finiteOutput(value);
  if (request.clamp === null) return result;
  if (request.clamp === "clamp") return Math.min(1, Math.max(0, result)) || 0;
  if (result < 0 || result > 1) {
    throw new ContractValidationError(ContractCode.OUT_OF_RANGE);
  }
  return result;
}

function point(
  request: AdditiveRequest,
  rows: ReadonlyArray<ExplainedComponent>,
): number {
  let total = finiteOutput(request.intercept);
  for (const item of rows)
    total = apply(total, item.contribution, item.operation);
  return final(total, request);
}

function advanceBounds(
  low: number,
  high: number,
  term: AdditiveTerm,
): readonly [number, number] {
  const [termLow, termHigh] = termBounds(term);
  return term.operation === "add"
    ? [apply(low, termLow, "add"), apply(high, termHigh, "add")]
    : [apply(low, termHigh, "subtract"), apply(high, termLow, "subtract")];
}

function resultInterval(request: AdditiveRequest): Interval | null {
  if (!request.terms.some((term) => term.interval !== null)) return null;
  let low = finiteOutput(request.intercept);
  let high = low;
  for (const term of request.terms)
    [low, high] = advanceBounds(low, high, term);
  const result = { low: final(low, request), high: final(high, request) };
  return result.low === result.high ? null : result;
}

export function additive(input: AdditiveRequest): ScoreResult {
  const request = parseRequest(input);
  if (request.method !== "additive") {
    throw new ContractValidationError(ContractCode.INVALID_METHOD);
  }
  const rows = request.terms.map(row);
  return parseScoreResult({
    schema: "assay.result/v1",
    method: { id: request.method, version: request.method_version },
    score: point(request, rows),
    interval: resultInterval(request),
    clamp: request.clamp,
    intercept: request.intercept,
    weight_total: null,
    components: rows,
    inputs_hash: inputsHash(request),
    selected_component_id: null,
  });
}

File: src/assay/additive.py
"""Declared-order additive and subtractive score composition."""

from __future__ import annotations

from typing import NoReturn

from assay.composite import canonical_zero, finite_output, inputs_hash, interval_or_none
from assay.contracts import (
    AdditiveRequest,
    AdditiveTerm,
    ClampPolicy,
    ExplainedComponent,
    Interval,
    Method,
    Operation,
    ScoreResult,
)
from assay.errors import ContractCode, ContractValidationError


def _fail(code: ContractCode) -> NoReturn:
    raise ContractValidationError(code) from None


def _contribution(term: AdditiveTerm, value: float | None = None) -> float:
    raw = term.value if value is None else value
    return finite_output(raw * term.coefficient)


def _apply(total: float, contribution: float, operation: Operation) -> float:
    if operation is Operation.ADD:
        return finite_output(total + contribution)
    return finite_output(total - contribution)


def _final(value: float, policy: ClampPolicy | None) -> float:
    result = finite_output(value)
    if policy is None:
        return result
    if policy is ClampPolicy.CLAMP:
        return canonical_zero(min(1.0, max(0.0, result)))
    if not 0.0 <= result <= 1.0:
        _fail(ContractCode.OUT_OF_RANGE)
    return result


def _explain(term: AdditiveTerm) -> ExplainedComponent:
    return ExplainedComponent(
        id=term.id,
        raw=term.value,
        normalized=None,
        operation=term.operation,
        coefficient=term.coefficient,
        contribution=_contribution(term),
        contribution_interval=interval_or_none(*_term_bounds(term)),
    )


def _point(request: AdditiveRequest, rows: tuple[ExplainedComponent, ...]) -> float:
    total = finite_output(request.intercept)
    for row in rows:
        total = _apply(total, row.contribution, row.operation)
    return _final(total, request.clamp)


def _term_bounds(term: AdditiveTerm) -> tuple[float, float]:
    if term.interval is None:
        contribution = _contribution(term)
        return contribution, contribution
    return _contribution(term, term.interval.low), _contribution(term, term.interval.high)


def _advance_bounds(low: float, high: float, term: AdditiveTerm) -> tuple[float, float]:
    term_low, term_high = _term_bounds(term)
    if term.operation is Operation.ADD:
        return finite_output(low + term_low), finite_output(high + term_high)
    return finite_output(low - term_high), finite_output(high - term_low)


def _result_interval(request: AdditiveRequest) -> Interval | None:
    if not any(term.interval is not None for term in request.terms):
        return None
    low = high = finite_output(request.intercept)
    for term in request.terms:
        low, high = _advance_bounds(low, high, term)
    return interval_or_none(_final(low, request.clamp), _final(high, request.clamp))


def additive(request: AdditiveRequest) -> ScoreResult:
    """Compose an explicit left-to-right sum with an optional final unit bound."""
    validated = AdditiveRequest.model_validate(request)
    rows = tuple(_explain(term) for term in validated.terms)
    return ScoreResult(
        method=Method(id=validated.method, version=validated.method_version),
        score=_point(validated, rows),
        interval=_result_interval(validated),
        clamp=validated.clamp,
        intercept=validated.intercept,
        components=rows,
        inputs_hash=inputs_hash(validated),
        selected_component_id=None,
    )
